ui/src/app/cohort-search/modal/modal.component.tsx

```diff
@@ -0,0 +1,597 @@
+  title: string;
+  treeSearchTerms: string;
+}
+
+export class CBModal extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      attributesNode: undefined,
+      autocompleteSelection: undefined,
+      backMode: undefined,
+      conceptType: null,
+      count: 0,
+      disableFinish: false,
+      groupSelections: [],
+      hierarchyNode: undefined,
+      loadingSubtree: false,
+      mode: 'list',
+      open: false,
+      selectedIds: [],
+      selections: [],
+      title: '',
+      treeSearchTerms: '',
+    };
+  }
+
+  componentDidMount(): void {
+    const {searchContext: {domain, item, standard, type}} = this.props;
+    if (!this.state.open) {
+      const selections = item.searchParameters;
+      const selectedIds = selections.map(s => s.parameterId);
+      if (type === CriteriaType.DECEASED) {
+        this.selectDeceased();
+      } else {
+        const title = domain === DomainType.PERSON ? typeToTitle(type) : domainToTitle(domain);
+        let backMode, hierarchyNode, mode;
+        if (this.initTree) {
+          hierarchyNode = {
+            domainId: domain,
+            type: type,
+            isStandard: standard,
+            id: 0,
+          };
+          backMode = 'tree';
+          mode = 'tree';
+        } else {
+          backMode = 'list';
+          mode = 'list';
+        }
+        this.setState({backMode, hierarchyNode, mode, open: true, selectedIds, selections, title});
+      }
+    }
+  }
+
+  setScroll = (id: string) => {
+    const nodeId = `node${id}`;
+    const node = document.getElementById(nodeId);
+    if (node) {
+      setTimeout(() => node.scrollIntoView({behavior: 'smooth', block: 'center'}), 200);
+    }
+    this.setState({loadingSubtree: false});
+  }
+
+  back = () => {
+    if (this.state.mode === 'tree') {
+      this.setState({autocompleteSelection: undefined, backMode: 'list', hierarchyNode: undefined, mode: 'list'});
+    } else {
+      this.setState({attributesNode: undefined, mode: this.state.backMode});
+    }
+  }
+
+  finish = () => {
+    const {searchContext: {domain, groupId, item, role, type}} = this.props;
+    const {selections} = this.state;
+    if (domain === DomainType.PERSON) {
+      triggerEvent('Cohort Builder Search', 'Click', `Demo - ${typeToTitle(type)} - Finish`);
+    }
+    const searchRequest = searchRequestStore.getValue();
+    item.searchParameters = selections;
+    if (groupId) {
+      const groupIndex = searchRequest[role].findIndex(grp => grp.id === groupId);
+      if (groupIndex > -1) {
+        const itemIndex = searchRequest[role][groupIndex].items.findIndex(it => it.id === item.id);
+        if (itemIndex > -1) {
+          searchRequest[role][groupIndex].items[itemIndex] = item;
+        } else {
+          searchRequest[role][groupIndex].items.push(item);
+        }
+      }
+    } else {
+      const group = initGroup(role, item);
+      searchRequest[role].push(group);
```

Comment:
nit: Totally optional, this can just be searchRequest[role].push(initGroup(role, item).

Reply:
I like that better too. Changed.